### PR TITLE
WIP: Adds TrainValidDataset

### PR DESCRIPTION
**Pull request by thomasjpfan**
This is PR will be updated based on the discussions in #282.

Currently, `TrainValidDataset` is a namedtuple containing a training and validation dataset. When `get_split_datasets` is called, it first checks for this namedtuple, and returns the corresponding training and validation datasets.

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Continuing the discussion from #282.

If this case is common enough, skorch should make room for it. I have to say, though, that I'm not 100% content with this solution. Sure, it might be necessary, but it clashes with many parts of skorch:

* What to do when you use this but also define a `train_split`? Having both makes no sense.
* When my data is not already in a dataset, I have to wrap it myself, which is normally something skorch does under the hood.
* This is difficult to use in an sklearn pipeline (do we need a helper transformer for that?).
* skorch already supports a similar way to achieve this outcome ([`PredefinedSplit`](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.PredefinedSplit.html)).

Of course, these kinds of conflicts will necessarily arise when building a bridge between two worlds, but I have the feeling we should encourage one best practice. For me, this is having my data in a more universal format (numpy arrays, dicts, etc.) and not in a custom data format that only works with pytorch and nowhere else.

Long story short: I agree with this PR (when some of the bullet points are fixed) because it facilitates working with pre-existing splits, but would always encourage to take the effort to have the data in a universal format as long as possible.

File: skorch/dataset.py [hunk 1]
```diff
@@ -281,3 +281,23 @@ def __call__(self, dataset, y=None, groups=None):
     def __repr__(self):
         # pylint: disable=useless-super-delegation
         return super(CVSplit, self).__repr__()
+
+
+class TrainValidDataset:
+    """Small wrapper class to support custom training and validation
```
Comment (BenjaminBossan) [inline skorch/dataset.py]: Here and further down, you describe this as "custom datasets" but this is misleading. A custom dataset suggests that I implemented my own type of dataset. I would rather go with "predefined" or "pre-existing split" (maybe the latter to avoid confusion with sklearn's `PredefinedSplit`). Also, I would add a newline after the first sentence.

File: skorch/net.py [hunk 2]
```diff
@@ -607,10 +608,14 @@ def fit_loop(self, X, y=None, epochs=None, **fit_params):
             * a dictionary of the former three
             * a list/tuple of the former three
             * a Dataset
+            * a TrainValidDataset
```
Comment (BenjaminBossan) [inline skorch/net.py]: maybe add: a skorch.dataset.TrainValidDataset, since most people won't know where this comes from.

File: skorch/tests/test_net.py [hunk 3]
```diff
@@ -1444,3 +1444,14 @@ def get_step(self):
         expected_losses = list(
             flatten(net.history[:, 'batches', :, 'train_loss']))
         assert np.allclose(side_effect[2::3], expected_losses)
+
+    def test_train_valid_dataset(self, net_cls, module_cls, data, dataset_cls):
```
Comment (BenjaminBossan) [inline skorch/tests/test_net.py]: What is missing somewhat from these tests is to show that `train_ds` is actually used for training and `valid_ds` for validation. This could be achieved, e.g., by mocking `get_iterator` and tracking what it is called with.
